[SIG-scale] Add density performance tests to evaluate the Kubevirt control plane

marceloamaral opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
This PR adds support for performance tests to be the basis of both performance and scalability tests. 

More specifically, It enables the creation of density tests to run (e.g., scheduling in parallel 100, 200 ... 600 VMs), measuring the VM's creation time (from submission to Running Phase) and failing when the time is higher than a given threshold. 

For further analysis, the VMI's phase transition latency, the VMI resource usage, and the Kubevirt control plane resource usage are also collected and logged into a file. Note that, resource usage is only collected when Prometheus is available.

The test configuration is defined in a YAML file so that it can be easily adapted to different environments. For rate control during the creation process, it is possible to define a step function or use a Poisson process with a given arrival rate.

The density test is inspired by the [Kubernetes density test](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/density_test.go). Even though  Kubernetes has a [new repository](https://github.com/kubernetes/perf-tests) for perf tests, the ginkgo-based tests are a good source of reference.


_why we need it_:
We would like to integrate those tests into our CI/CD system to facilitate the evaluation of the Kubevirt control plane. The motivation of such evaluation is because code changes may affect system scalability and performance. Then, with a set of performance and scalability tests, we may identify possible performance regression in the KubeVirt control plane between PRs and new Releases.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
The thresholds now are pessimistic since my environment has nested virtualization and I could not measure "realistic" latencies yet. I will need to adapt it accordingly...

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

marceloamaral commented:
/retest

rthallisey commented on tests/performance/metrics/metrics.go:
```diff
@@ -0,0 +1,212 @@
+
+// ExtractMetrics returns latency metrics for each percentile(50th, 90th and 99th).
+func ExtractMetrics(data []VMMetricData) Metric {
+	metrics := Metric{}
```
Can this leverage the work going on in #5766 ? 5766 calculated the transitions times so you can scrape the endpoint.

marceloamaral replied:
Well, yes and no... The problem to be strictly dependent on the Prometheus metric is that the test can only run if Prometheus is deployed. Which is not done by default in `kubevirtci`, although, I have enabled the optinal deployment of Prometheus in `kubevirtci` in [#608](https://github.com/kubevirt/kubevirtci/pull/608).

Prometheus could potentially be a strong requirement for us, but, I intended to have this test as flexible as possible so that everyone could execute it into their environment.

Note that, for the perspective of our CI/CD system, the metrics from #5766 will be very valuable to visualize the tests later in a Grafana dashboard.

rmohr replied:
I guess that is something we have to discuss further. It would be weird to use one set of metrics to decide if we meet certain goals but we use different data to visualize the data.

rthallisey commented on tests/performance/utils.go:
```diff
@@ -0,0 +1,145 @@
+}
+
+// DumpDataToFile inserts the current timestamp into the labels and writes the data for the test into the file with the specified sufix.
+func dumpDataToFile(data interface{}, labels map[string]string, sufix string) {
```
Performance data output seems like a good abstraction.  In other words, if we capture any data for perf or scale, the output can end up in prometheus, stdout, to a file, ect..., depending on how a user configures kubevirt.  Can we scope out an interface that would work for all use cases?

marceloamaral replied:
I tried to create the `PerfData` and `DataItem` structures very generic, so I believe it can support different cases.

```
// DataItem is the data point.
type DataItem struct {
	Data map[string]float64 `json:"data"`
	Unit string `json:"unit"`
	Labels map[string]string `json:"labels,omitempty"`
}

// PerfData contains all data items generated in current test.
type PerfData struct {
	Version   string     `json:"version"`
	DataItems []DataItem `json:"dataItems"`
	Labels map[string]string `json:"labels,omitempty"`
}
```

The hard constraint is that the values must always be `float64`, but the unit and names (i.e., labels) are any string value.

marceloamaral replied:
Btw, [here](https://docs.google.com/document/d/1g5tjv1GcxNsw0pDmToz6z5NRj9dVZTnbsAOvjj8Qdek/edit) is an example of the output of this test.

rthallisey commented on tests/performance/resourcecollector/resource_collector.go:
```diff
@@ -0,0 +1,282 @@
+}
+
+// GetApiRequestDuration queries Prometheus API endpoint for the kubevirt API REST request duration, it returns the 99-percentile of the latencies of APIS
+func (rc *ResourceCollector) GetApiRequestDuration(url string) (prom_model.Value, error) {
```
We definitely want to have this information exported, but I don't see you wiring up many of these functions to anything.  Do these helper functions need to be a part of this PR?

marceloamaral replied:
Humm, it is a good point! I forgot to include an API request latency verification...

The plan is to actually verify the API latency and fail the test if the latency is much higher than expected.
Also, note that, I am verifying only the calls related to VMI creation, there are other calls that I am ignoring for now...

So, I could remove this, or include the latency verification that is missing....

rthallisey replied:
I think there are a few ideas encapsulated in this PR and each of them are important.  I'd like to break down what I think you are trying to accomplish into individual designs so each idea gets the attention it deserves.  Here's my understanding:
1) Create a test framework that captures perf thresholds and reports the results to std out.  The current target looks to be for CI but maybe it can be used by developers testing code locally too?
2) Collect metrics for Kubernetes API calls - https://github.com/kubevirt/kubevirt/issues/5732
3) Collect metrics for VM pod mem/cpu - https://github.com/kubevirt/kubevirt/issues/5732
4) Define an API for different types of reporting

Maybe we can start by creating some issue for 1) and 4)

marceloamaral replied:
>1. Create a test framework that captures perf thresholds and reports the results to std out. The current target looks to be for CI but maybe it can be used by developers testing code locally too?

That is right. Note that, for the CI standpoint, the test will fail if the latency is not satisfied.

marceloamaral replied:
> 2. Collect metrics for Kubernetes API calls - #5732

Actually, the test focus on collecting the latency of KubeVirt API calls
For example, the following command returns the latency of the following APIs:
```
histogram_quantile(0.99, 
    sum(rest_client_request_duration_seconds_bucket{url=~".*kubevirt.*"}) by(url,verb,le)
)
```
```
{url="https://virt-api.kubevirt.svc:443/virtualmachineinstances-mutate;verb="POST"}
{url="https://kubevirt-operator-webhook.kubevirt.svc:443/kubevirt-validate-update;verb="POST"}
{url="https://virt-api.kubevirt.svc:443/virtualmachineinstances-validate-update;verb="POST"}
{url="https://kubevirt-operator-webhook.kubevirt.svc:443/kubevirt-validate-delete;verb="POST"}
{url="https://virt-api.kubevirt.svc:443/virtualmachineinstances-validate-create;verb="POST"}
```
But, of course, without the `url` filter, we can collect for all Kubernetes APIs as well... But, this is not the goal of this PR.

marceloamaral replied:
> 3. Collect metrics for VM pod mem/cpu - #5732

Yes. 

Also, in my test, I saw a big CPU overhead. 
The aggregated CPU usage in the VMI namespace was the double of the sum of all VMI CPU limits.

marceloamaral replied:
> 4. Define an API for different types of reporting

Right, I see this as a consequence of implementing the test.

rthallisey replied:
>        Collect metrics for Kubernetes API calls - #5732
> Actually, the test focus on collecting the latency of KubeVirt API calls
> For example, the following command returns the latency of the following APIs:
>
I'll add api latency to #5732.

rthallisey commented:
cc @davidvossel @rmohr

marceloamaral commented on tests/performance/density.go:
```diff
@@ -0,0 +1,471 @@
+	Expect(err).ToNot(HaveOccurred())
+}
+
+func attemptToLoginToVM(virtClient kubecli.KubevirtClient, testArg DensityTest, vmi *apiv1.VirtualMachineInstance) *apiv1.VirtualMachineInstance {
```
This part I am not sure if I should keep in the PR.

Right now, I agreed with @rmohr that the test in our CI/CD system will not boot the VMs to be able to scale it and put more pressure in the control plane. The creation time is the time interval from submission to Running Phase.

However, I made the test as flexible as possible that someone could easily change the VM creation time from submission up to when we can ping the VM. To be able to test the whole VM creation pipeline, and make sure that all components are properly working (e.g. the network)...

rthallisey replied:
I spoke with @davidvossel previously about an idea I had that's similar.  I'll share the WIP design document at the next SIG-scale and we should discuss this idea as well.

marceloamaral replied:
Sure! Thanks!
Btw, we could start this performance test with the approach that I am suggesting in this PR, and extend it in the future. wdyt?

marceloamaral commented:
cc @nelsonspbr @ezrasilvera

marceloamaral commented:
It seems that we all agreed to approve this PR.
/approve

marceloamaral commented:
Sorry, I sent the approve message in the wrong PR....

nelsonspbr commented:
cc @fabiand @dhiller @jean-edouard @fgimenez @vladikr @omeryahud

nelsonspbr commented:
@marceloamaral Might need to add a release note, even if it is just `NONE`, to allow this to be merged in the future.

nelsonspbr commented:
@marceloamaral I saw you linked a doc above ([here](https://docs.google.com/document/d/1g5tjv1GcxNsw0pDmToz6z5NRj9dVZTnbsAOvjj8Qdek/edit)). Is it a direct result of the implementation you are proposing in this PR? I think it is a great set of initial results and I think we could easily expand it to many more VMs by simply adding a few more nodes. I can help you with that if you need.

marceloamaral commented:
@nelsonspbr yes, this document shows an example of the outputs of the test in this PR

rthallisey commented:
https://github.com/kubevirt/kubevirt/issues/5878

davidvossel reviewed (commented):
I like the idea of a repeatable way to create load with some sort of batch VMI create density test, and I like the idea of having code that can monitor performance during this density test.

What I'd like even more is if these two things weren't coupled together.

What would you think about leaving the density test pretty much as you have it, but abstracting away the logic that is meansuring performance into a standalone tool, perhaps called `performance-profiler` which could live in the kubevirt/kubevirt repo in `tools/performance-profiler/performance-profiler.go`

The performance-profiler would discover the control plane and VMIs as they are introduced, then profile their performance until the process receives some sort of stop signal. 

What this would allow is for us to create a variety of density and performance scenarios independently of our test suite. I could for example use a standalone k8s load generator like `kube-burner` to generate the VMIs and specific intervals, and then use the `performance-profiler` to provide the results in a structured format.

marceloamaral commented:
> What would you think about leaving the density test pretty much as you have it, but abstracting away the logic that is meansuring performance into a standalone tool, perhaps called `performance-profiler` which could live in the kubevirt/kubevirt repo in `tools/performance-profiler/performance-profiler.go`
> 
> The performance-profiler would discover the control plane and VMIs as they are introduced, then profile their performance until the process receives some sort of stop signal.
> 

@davidvossel I think it make sense, let's discuss more about the design...

In this case, the `performance-profiler` would be a binary that we start independently from the test....
I don't want to make it too complex, but it might become.... and maybe takes time to implement....
Any way, let's list what the `performance-profiler` should have for the density test standpoint:
- [ ] watch for VMs, might be an operator then....wdyt?
- [ ] track the VM phases independently from Prometheus...(maybe not since we have the histograms now)
- [ ] store the VM's information, VM phases and VM pod resource usage. We can extend the metrics later as needed...
- [ ] an API to expose the results, start, stop the collection, etc...
- [ ] reports results to stdout and optionally into a file

fgimenez commented:
@marceloamaral WRT the failing `pull-kubevirt-checkunassigned-tests` we need to add `[sig-performance]` to https://github.com/kubevirt/kubevirt/blob/main/hack/check-unassigned-tests.sh#L4

marceloamaral commented:
@rthallisey @rmohr and @vossel I updated the PR as we discussed last time.

The PR was simplified, and all the metric collection part from Prometheus was removed. The flexible configuration from a YAML was also removed. As we agreed, I kept only the functional test.

Could you please review this PR?

rmohr reviewed (commented):
I was hoping for something much more basic initially which just creates the load and make the test data identifyable in prometheus.

Basically something simple like this:

```go
	BeforeEach(func(){
		tests.BeforeTestCleanup()
		// wait a little to ensure that we get rid of some general cluster noise from the cleanup
		time.Sleep(30* time.Second)
		virtClient, err = kubecli.GetKubevirtClient()
		util.PanicOnError(err)
	})

	It("should do a scale test", func(){
	        // Change the global prometheus configuration to add this label to all metrics
		AddTestLabel("testidentifyinglabel")
		// wait a little to ensure that prometheus picks the change up
		time.Sleep(30* time.Second)
		templateVMI := myStandardNotBootingCirrosWithGenerateName()
		stopCh := make(chan struct{})
		created := 0
		// create a constant ticker which will create VMs at a constant rate (non sliding)
		wait.JitterUntil(func() {
			if created == 0 {
				close(stopCh)
				return
			}
			_, err := virtClient.VirtualMachineInstance(util.NamespaceTestDefault).Create(templateVMI)
			Expect(err).ToNot(HaveOccurred())
			created++
		}, 100*time.Millisecond, 0.0, false, stopCh)

		Eventually(func() int {
			vmis, err := virtClient.VirtualMachineInstance(util.NamespaceTestDefault).List(&metav1.ListOptions{})
			Expect(err).ToNot(HaveOccurred())
			running := 0
			for _, vmi := range vmis.Items {
				if vmi.Status.Phase == v1.Running {
					running++
				}
			}
			return running
		}, 180 * time.Second, 10 * time.Second).Should(Equal(100))
	})


	AfterEach(func(){
		// ensure the metrics get scraped and collected  till the end
		time.Sleep(30* time.Second)
		RemoveTestLabel()
		// wait a little to ensure that prometheus picks the change up
		time.Sleep(30* time.Second)
	})
```

At least initially I would hope to leverage separating creating the load and collecting the data completely and just look at prometheus what we get. At this stage I would want to avoid having to interpret different datapoints from different collection points (tests doing their own thing and prometheus doing its own thing).

@marceloamaral @davidvossel @rthallisey  what do you think?

marceloamaral commented on vendor/github.com/prometheus/client_golang/api/client.go:
```diff
@@ -0,0 +1,129 @@
+// Copyright 2015 The Prometheus Authors
```
Yes, I will update that

marceloamaral replied:
> AddTestLabel("testidentifyinglabel")

@rmohr unfortunately, I did not find a way to add a global label to all metrics from the application. 
Do you know how to do that?

rmohr replied:
That should work with a `RelabelConfig`: https://github.com/prometheus-operator/prometheus-operator/blob/master/Documentation/api.md.

Something like this should do the trick: 

```yaml
sourceLabels:
  - "__address__"
targetLabel: "test"
replacement: "<testid>"
```

Every metric has a `__address__` field, therefore every metric which we collect during that time should be matched (Here more explanations: https://albersdevelopment.net/2019/08/28/prometheus-adding-a-label-to-a-target/).

marceloamaral commented on tests/performance/density.go:
```diff
@@ -0,0 +1,323 @@
+
+	tests.WaitForSuccessfulVMIStartWithTimeoutIgnoreWarnings(vmi, int(testArg.VMStartupLimits.Perc99))
+
+	vmi, err = virtClient.VirtualMachineInstance(vmi.Namespace).Get(vmi.Name, &metav1.GetOptions{})
```
It makes sense, I updated that.
However, for deletion we cannot watch since we need to verify if the object has disappeared.

rthallisey replied:
You will see a Deleting event in the watch, so you can limit your polling to start only after the Deleting event is seen.

marceloamaral commented:
@rthallisey I updated the test accordingly to your comments.
@rmohr I am not sure if we need to identify the test in Prometheus since the Prow job will do that.

Regarding collecting the data, I am using the same timestamps as the prometheus metrics are using. It must not be different...
Only with the VMI creation time, we can set the test thresholds...

rmohr commented:
> @rmohr I am not sure if we need to identify the test in Prometheus since the Prow job will do that.

For a single test it is not relevant. I was more thinking about other tests too, so that we can clearly see results per test.

rmohr commented:
> Regarding collecting the data, I am using the same timestamps as the prometheus metrics are using. It must not be different...
> Only with the VMI creation time, we can set the test thresholds...

Well, you are watching. They are probably similar. I would prefer to stick with prometheus. If possible, anything which we need to see issues in our cluster for scale testing would be great to have available and as the primary source also in production.

rmohr commented:
> Well, you are watching. They are probably similar if there are no errors, but not the same. I would prefer to stick with prometheus. If possible, anything which we need to see issues in our cluster for scale testing would be great to have available and as the primary source also in production.

And I am fine with creating a freamework and all, and also doing some measurements outside of prometheus in general, but let's really start with a useful bare minimum.

VirrageS commented on tests/performance/density.go:
```diff
@@ -0,0 +1,316 @@
+	ArrivalRateFuncName  string            `yaml:"arrival_rate_func_name"`  // e.g. using a constant interval or a poisson process
+	VMImage              cd.ContainerDisk  `yaml:"vm_image_name"`           // the image type will reflect in the storage size
+	CPULimit             string            `yaml:"cpu_limit"`               // number of CPUs to allocate (100m = .1 cores)
+	MEMLimit             string            `yaml:"mem_limit"`               // amount of Memory to allocate
```
```suggestion
	MemLimit             string            `yaml:"mem_limit"`               // amount of Memory to allocate
```

VirrageS commented on tests/performance/utils.go:
```diff
@@ -0,0 +1,91 @@
+	}
+
+	randNumber := r.Float64()
+	inter_arrival_time := -math.Log(1.0-randNumber) / float64(lambda)
```
```suggestion
	interArrivalTime := -math.Log(1.0-randNumber) / float64(lambda)
```

marceloamaral replied:
@rmohr  has suggested to remove the arrival process for now to simply the PR. I am removing that
Although, I think that creating VMs using a Poisson process is more realistic, we can include that in the future...

rmohr replied:
@marceloamaral thanks, following up the discussion on the community meeting. I suggested to remove everything which goes beyond just starting X vms. Including all structs where you want to use different functions or values. Just a plain test. Also all watching code and so forth. As @davidvossel, you and I discussed there ,we want to start completely without a framework and just a simple scaling test, starting collecting data in CI.

marceloamaral replied:
Thanks @rmohr! Right, but I think we agreed to have the thresholds. @rthallisey has written that in the meeting minutes.
To have the thresholds we need to watch the VMIs since we are not including Prometheus. Define the test parameters in a struct is to better organize and visualize the test. IMHO, the PR is not a framework, it is only a test that verifies the VM creation time...

rmohr replied:
@marceloamaral this is probably kind of a misunderstanding, but we argued a lot about separating out all interpretation and thresholds from this PR. I think we all agreed only that we want this in general at some point.

marceloamaral replied:
@rmohr as we discussed before, I removed all the latency collection, verification and configuration part of the PR.

I only kept the vm watcher to count the number of running VM. As pointed before, Listing operation suffers from API throttling issues, so I kept the watch. But I can also remove that if needed...

Could you please review it again? Let me know if there is still some part of the PR that should be removed...

davidvossel replied:
This #6035 kind of tool is what I'm hoping we can begin using for gathering and comparing performance results during these stress tests.

VirrageS commented on tests/performance/metrics/metrics.go:
```diff
@@ -0,0 +1,162 @@
+
+// Metric is a struct for metrics for performance tests, it can be latencies or resource usage.
+type Metric struct {
+	Perc50 float64 `yaml:"perc50"`
```
I would suggest making this a map: `Perc map[int]float64`. Adding or removing fields could be disruptive in the future. This would make this more general and we could easily add or remove values whenever we would want.

We could also have something like `Percentiles := []int{50, 90, 99}` which would be a static array of the percentiles that we would like to collect. Then setting `Metric.Perc` map values would be just a simple as:
```golang
for _, percentile := Percentiles {
  metric.Perc[percentile] = data[int(math.Ceil(float64(length*percentile)/100))-1].Value
}
```

marceloamaral replied:
It makes sense! Thanks

VirrageS commented on tests/performance/metrics/metrics.go:
```diff
@@ -0,0 +1,162 @@
+type BatchStatus struct {
+	sync.RWMutex
+	metrics map[kvv1.VirtualMachineInstancePhase]Slice
+	vmiMap  map[kvv1.VirtualMachineInstancePhase]map[string]kvv1.VirtualMachineInstance
```
I would suggest removing `Map` from the name. It is a `map` but it doesn't give us any more information about what exactly it keeps. Name suggestions:
* `vmis`
* `vmisByPhases`

marceloamaral replied:
Good point! I liked `vmisByPhases`

VirrageS commented on tests/performance/metrics/metrics.go:
```diff
@@ -0,0 +1,162 @@
+	defer s.Unlock()
+
+	if _, ok := s.vmiMap[kvv1.Running][vmi.ObjectMeta.Name]; !ok {
+
```
Sorry for being pedantic. But I don't fully understand why you sometimes have an empty line after the `if` statement and sometimes don't. Is there something I'm missing?

marceloamaral replied:
Actually, thanks for pointing it. There is not reason to have those line breaks!

VirrageS commented on tests/performance/density.go:
```diff
@@ -0,0 +1,316 @@
+	if t.ArrivalRate > 0 {
+		return time.Duration(1/t.ArrivalRate) * time.Millisecond
+	}
+	return time.Duration(300) * time.Millisecond
```
```suggestion
	return 300 * time.Millisecond
```

marceloamaral replied:
As I mentioned before, I am removing the Poisson Process

VirrageS commented on tests/performance/density.go:
```diff
@@ -0,0 +1,316 @@
+	densityTests := []DensityTest{
+		{
+			NumVMs:              100,
+			ArrivalRate:         0.002, // Vms per second
```
I don't fully understand this. Why `per second` if the `ArrivalUnit` is set to `time.Millisecond`?

VirrageS commented on tests/performance/density.go:
```diff
@@ -0,0 +1,316 @@
+	}
+	// ArrivalRate is the number fo vms per second. Then, we calculate a stepped interval between VMs
+	if t.ArrivalRate > 0 {
+		return time.Duration(1/t.ArrivalRate) * time.Millisecond
```
Shouldn't it be:
```suggestion
		return time.Duration(1/t.ArrivalRate) * t.ArrivalUnit
```

VirrageS commented on tests/performance/density.go:
```diff
@@ -0,0 +1,316 @@
+)
+
+type DensityTest struct {
+	NumVMs               int               `yaml:"num_vms"`                 // number of vms
```
```suggestion
	VMCount               int               `yaml:"vm_count"`                 // number of vms
```
I think it's often suggested to use `count` instead of `number` or `numberOf`: https://stackoverflow.com/questions/6358588/how-to-name-a-variable-numitems-or-itemcount

marceloamaral replied:
I will update to count

VirrageS commented on tests/performance/density.go:
```diff
@@ -0,0 +1,316 @@
+		gracePeriodSeconds := int64(0)
+		err := virtClient.VirtualMachineInstance(data.Namespace).Delete(data.Name, &metav1.DeleteOptions{GracePeriodSeconds: &gracePeriodSeconds})
+		Expect(err).To(BeNil())
+		time.Sleep(200 * time.Millisecond)
```
What is this sleep for?

marceloamaral replied:
The sleep here is to avoid to overload the Kubernetes API server with too much requests and become the bottleneck of the analysis.  Additionally, it also prevent the k8s client go to apply rate limiting (throttling) in the requests.

VirrageS replied:
Makes sense, thanks for the explanation. Maybe we could add comment here: `// Sleep to manually throttle API requests.`

VirrageS commented on tests/performance/density.go:
```diff
@@ -0,0 +1,316 @@
+	status.RUnlock()
+}
+
+func waitVMIBatchStatus(status *e2emetrics.BatchStatus, phase kvv1.VirtualMachineInstancePhase, numVMI int, timeout time.Duration) error {
```
```suggestion
func waitVMIBatchStatus(status *e2emetrics.BatchStatus, phase kvv1.VirtualMachineInstancePhase, vmiCount int, timeout time.Duration) error {
```

VirrageS commented on tests/performance/framework.go:
```diff
@@ -0,0 +1,55 @@
+
+func init() {
+	flag.BoolVar(&RunPerfTests, "performance-test", false, "run performance tests. If false, all performance tests will be skiped.")
+	ptest := os.Getenv("KUBEVIRT_E2E_PERF_TEST")
```
```suggestion
	ptest, _ := strconv.ParseBool(os.Getenv("KUBEVIRT_E2E_PERF_TEST"))
```

marceloamaral replied:
Much better approach! Thanks!

marceloamaral commented:
It was closed by mistake

dhiller commented:
/reopen

marceloamaral commented:
/retest

marceloamaral commented:
/retest

rmohr commented on tests/performance/density.go:
```diff
@@ -0,0 +1,186 @@
+		Expect(err).ToNot(HaveOccurred())
+
+		// sleep to manually throttle API requests.
+		time.Sleep(300 * time.Millisecond)
```
Do you try to protect  the API with that? If only that is the use-case, the value seems to be very high.

marceloamaral replied:
Right, the legend is not accurate. This sleep is to control the creation rate.
This is the same as in the Kubernetes density test [here](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/density_test.go#L349)

marceloamaral commented on tests/performance/density.go:
```diff
@@ -0,0 +1,186 @@
+
+			It("should sucessfully create all VMIS", func() {
+				quit := make(chan bool)
+				runningVMI := make(map[string]int)
```
Right, I had Locks before, I oversimplified this. Anyway, I will remove this part as you suggested...

rmohr reviewed (commented):
As discussed in the scale meeting the prometheus label can be added in a follow-up.

/lgtm
/approve

ormergi commented:
All the tests on SRIOV lane were skipped
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/5835/pull-kubevirt-e2e-kind-1.19-sriov/1415867369615527936

ormergi commented:
Hey @marceloamaral @rmohr, all tests on SRIOV lane were skipped
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/5835/pull-kubevirt-e2e-kind-1.19-sriov/1415867369615527936

stu-gott commented:
@marceloamaral apologies for the confusion. This PR was reverted because of CI impact. Would you be able to re-submit this PR with the problem corrected?

We apologize for the inconvenience. This certainly is an exceptional case.

rmohr commented on tests/performance/framework.go:
```diff
@@ -0,0 +1,55 @@
+}
+
+func SIGDescribe(text string, body func()) bool {
+	BeforeEach(func() {
```
This BeforeEach can't be here. That is what I missed in my earlier review. The BeforeEach is registered outside of the `Describe` below which makes it a global BeforeEach.

marceloamaral replied:
Thanks @rmohr. Thanks is right, I must move it to the test Description

rmohr commented on tests/performance/framework.go:
```diff
@@ -0,0 +1,55 @@
+}
+
+func FSIGDescribe(text string, body func()) bool {
+	BeforeEach(func() {
```
Same here.

rmohr commented on tests/performance/framework.go:
```diff
@@ -0,0 +1,55 @@
+var RunPerfTests = false
+
+func init() {
+	flag.BoolVar(&RunPerfTests, "performance-test", false, "run performance tests. If false, all performance tests will be skiped.")
```
Instead of adding this flag, can you for now just always skip `sig-performance` in `hack/functest.sh`? We can in a follow-up probably give the scale tests their own main entrypoint, but it should be good enough for now.

rmohr reviewed (commented):
@marceloamaral left some hints on what I missed on my last review.
